### DISCUSSION TITLE
booting from virtiofs fails with error `virtio-fs: tag </dev/root> not found`

### DISCUSSION THREAD

**Discussion by muhamadazmy** [cloud-hypervisor/cloud-hypervisor#2719] [General]
Host: Arch 5.12.8-arch1-1
Cloud-Hypervisor: v15.0.0

Cmdline:
```
sudo cloud-hypervisor \
   --kernel /boot/vmlinuz-linux  \
   --initramfs initramfs-linux.img  \
   --console off \
   --serial tty \
   --cpus boot=1 \
   --memory size=1024M,shared=on \
   --fs tag=/dev/root,socket=/tmp/mnt.socket  \
   --cmdline 'console=ttyS0 init=/usr/bin/redis-server rootfstype=virtiofs root=/dev/root rw' \
   --rng
```

Virtiofsd Cmdline:
```
sudo virtiofsd-rs --socket /tmp/mnt.socket --shared-dir /home/azmy/redis
```

## Issue
I am trying to boot a machine using a virtiofs `root`. I am using the same kernel as my host, but the initrafm imagine is custom build to include als the virtiofs module 

I end up dropping to the emergency shell

```
Waiting 10 seconds for device /dev/root ...
:: mounting '/dev/root' on real root
mount: /new_root: wrong fs type, bad option, bad superblock on /dev/root, missing codepage or helper program, or other error.
You are now being dropped into an emergency shell.
```

when I do dmesg i get
```
[   10.995937] virtio-fs: tag </dev/root> not found
```

trying to do some debugging:
```
[rootfs ]# lsmod 
Module                  Size  Used by
virtiofs               32768  0
fuse                  159744  1 virtiofs
crc32c_intel           24576  0
[rootfs ]# 
[rootfs ]# mount -t virtiofs /dev/root root
mount: /root: wrong fs type, bad option, bad superblock on /dev/root, missing codepage or helper program, or other error.
[rootfs ]# dmesg |tail
[    0.306077] Run /init as init process
[    0.306078]   with arguments:
[    0.306079]     /init
[    0.306079]   with environment:
[    0.306080]     HOME=/
[    0.306080]     TERM=linux
[    0.457341] fuse: init (API version 7.33)
[   10.995937] virtio-fs: tag </dev/root> not found
[  100.756807] random: fast init done
[  179.300654] virtio-fs: tag </dev/root> not found
```
Note the last log line  was trigger by the mount command.

I conclude from this that the module is available and loaded correctly, but the guest still can not mount the named tag. I also trying to change the tag name with no luck, the `/dev/root` was mentioned in this [doc](https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/docs/virtiofs-root.md) 

Do you think we require other modules, and/or binaries to be included in the initramfs ? is there a way to debug what is being exposed by the virtiofs on the guest ?

**Comment by sboeuf** [comment]
Can you try with the C version `virtiofsd` instead of the Rust one `virtiofsd-rs`?
Also, make sure you can boot a regular Cloud Hypervisor VM with `virtio-fs` first, and check that you can mount and access the content of the virtio-fs shared directory. This would be a good step to make sure you have virtio-fs working.
And last thing, make sure you run the most recent kernel version in the guest, so that you rely on the latest virtio-fs patches for the driver.

**Reply by muhamadazmy** [reply]
I already tried the C version `virtiofsd` with exactly same result. 

I will try another image with virtio-fs, and let u know.

**Comment by muhamadazmy** [comment]
I did few tests, and changed how we build the initramfs. It worked when I made sure that I include ALL modules for my kernel. after the boot, I only reduced the list to the set of module that where actually in use, and this ended up like:

```
crc32_pclmul           16384  0
crc32c_intel           24576  0
virtiofs               32768  1
fuse                  159744  1 virtiofs
virtio_blk             20480  0
aesni_intel           376832  0
virtio_rng             16384  0
rng_core               16384  1 virtio_rng
crypto_simd            16384  1 aesni_intel
virtio_pci             24576  0
cryptd                 24576  1 crypto_simd
virtio_pci_modern_dev    16384  1 virtio_pci
```

**Reply by sboeuf** [reply]
So just to clarify, did you fix the issue you reported before?

**Reply by muhamadazmy** [reply]
Yes, thanks for your help, and sorry about your time :)

**Reply by sboeuf** [reply]
Oh no worries :)
I'm glad you got your issue fixed!
